p2p/host.go

```diff
@@ -90,6 +99,7 @@ type Config struct {
 	DisableDHT                  bool        `mapstructure:"disable-dht"`
 	Flood                       bool        `mapstructure:"flood"`
 	Listen                      string      `mapstructure:"listen"`
+	ListenAddresses             []string    `mapstructure:"listen-addresses"`
```

Comment:
Yes, I need to make config validation fail if this happens. Will fix

Reply:
Upon further consideration, found out (or, to be more precise, remembered) that the problem cannot be fixed that easily due to backward compatibility concerns. Namely, previously we had ListenAddress (`listen-address`) which was just a string, and now we have ListenAddresses (`listen-addresses`) which is a list of strings. The value must default to `/ip4/0.0.0.0/tcp/7513`. My idea was to produce an error when both `listen-address` and `listen-addresses` are set, but the problem is, if one of these fields has a default value, the validation will always fail when the other is specified. So, I had to resort to just ignoring `listen-addresses` when `listen-address` is specified, which is far from perfect, but at least is backward-compatible. So we either need to keep this scheme or alter the defaulting mechanism such as having no default for the config field and using default `ListenAddresses` only when configuring the host.

Re `advertise-addresses`, we could handle it "better", but that would be somewhat inconsistent

Reply:
We want to force users to update their config to the new fields and print an error during startup if they still use an old (and now invalid) config.

To achieve this you can do the following:

- specify a custom type `DeprecatedListener` (or similar) that is a `string` but also implements the `pflag.Value` interface
- in the `Set` method of this interface you can return an `error` and you need to update `EnsureCLIFlags` in `base.go` with an additional case with your new type that also panics when the value is tried to be set
- for examples of custom types in our config you can see `activation.PowDifficulty` or `activation.PostProviderID`

The error that you return or the panic you throw should indicate to the user that the config field is not valid any more and that they have to specify the new field instead. Also add that information to `CHANGELOG.md`.

This way users are forced to update their config to the new schema. We can remove the old (now invalid/unused fields) in a future version then.

Reply:
I like this idea but what I'm afraid here is impact on our own infra, in particular. This might have to be adjusted
in multiple places and that might cause some extra hassle; I will ask other folks about their opinion on this, too.

But that actually gave me a possible idea how to deal with it in a backward-compatible way:
* define a new type that is `[]string` and can be decoded from both a list of strings and just a string by mapstructure, e.g. `type StringList []string`
* instead of adding new fields, just make `Listen` and `AdvertiseAddress` fields to be `StringList`. This way, old configs will still be handled, and `--listen` and `--advertise-address` options will still work, and you can also pass multiple `--listen` and `--advertise-address` options to specify multiple addrs.

We can optionally log warnings about `StringList` being decoded from simple string in the config, and later remove this possibility, should that be needed.

The drawback of this approach is slight misnomer, b/c in the config we'll have `advertise-address` which can be a list of strings instead of just a single string. For `listen` I think it is ok.

Reply:
Implemented in backward-compatible way like described, but using
```golang
type AddressList []multiaddr.Multiaddr
```